src/main/java/org/traccar/geocoder/AmapGeocoder.java

```diff
@@ -0,0 +1,137 @@
+                address.setSettlement(getJsonNode(addressComponent, "district"));
+                address.setDistrict(getJsonNode(addressComponent, "city"));
+                address.setState(getJsonNode(addressComponent, "province"));
+                address.setCountry(getJsonNode(addressComponent, "country"));
+            }
+            return address;
+        }
+        return null;
+    }
+
+    private String getJsonNode(JsonObject json, String key) {
+        if (json != null && json.containsKey(key)) {
+            JsonValue value = json.get(key);
+            if (value.getValueType() == JsonValue.ValueType.STRING) {
+                return ((JsonString)value).getString();
+            }
+        }
+        return null;
+    }
+
+    protected String parseError(JsonObject json) {
+        return json.getString("info");
+    }
+
+    @Override
+    public String getAddress(
+            final double latitude, final double longitude, final ReverseGeocoderCallback callback) {
+        if (outOfChina(latitude, longitude)) {
+            return null;
+        }
+        double dLat = transformLat(longitude - 105.0, latitude - 35.0);
+        double dLon = transformLon(longitude - 105.0, latitude - 35.0);
+        double radLat = latitude / 180.0 * pi;
+        double magic = Math.sin(radLat);
+        magic = 1 - ee * magic * magic;
+        double sqrtMagic = Math.sqrt(magic);
+        dLat = (dLat * 180.0) / ((a * (1 - ee)) / (magic * sqrtMagic) * pi);
+        dLon = (dLon * 180.0) / (a / sqrtMagic * Math.cos(radLat) * pi);
+        double mgLat = latitude + dLat;
+        double mgLon = longitude + dLon;
+
+        return super.getAddress(mgLat, mgLon, callback);
+    }
+
+    private static boolean outOfChina(double lat, double lon) {
+        if (lon < 72.004 || lon > 137.8347)
+            return true;
+        if (lat < 0.8293 || lat > 55.8271)
+            return true;
+        return false;
+    }
+
+    private static double transformLat(double x, double y) {
+        double ret = -100.0 + 2.0 * x + 3.0 * y + 0.2 * y * y + 0.1 * x * y
+                + 0.2 * Math.sqrt(Math.abs(x));
+        ret += (20.0 * Math.sin(6.0 * x * pi) + 20.0 * Math.sin(2.0 * x * pi)) * 2.0 / 3.0;
+        ret += (20.0 * Math.sin(y * pi) + 40.0 * Math.sin(y / 3.0 * pi)) * 2.0 / 3.0;
+        ret += (160.0 * Math.sin(y / 12.0 * pi) + 320 * Math.sin(y * pi / 30.0)) * 2.0 / 3.0;
+        return ret;
+    }
+
+    private static double transformLon(double x, double y) {
+        double ret = 300.0 + x + 2.0 * y + 0.1 * x * x + 0.1 * x * y + 0.1
+                * Math.sqrt(Math.abs(x));
+        ret += (20.0 * Math.sin(6.0 * x * pi) + 20.0 * Math.sin(2.0 * x * pi)) * 2.0 / 3.0;
+        ret += (20.0 * Math.sin(x * pi) + 40.0 * Math.sin(x / 3.0 * pi)) * 2.0 / 3.0;
+        ret += (150.0 * Math.sin(x / 12.0 * pi) + 300.0 * Math.sin(x / 30.0
+                * pi)) * 2.0 / 3.0;
+        return ret;
```

Comment:
In China, map providers are required by law to use the GCJ-02 geographic coordinate system (https://en.wikipedia.org/wiki/Restrictions_on_geographic_data_in_China), and the code here is used to convert the WG-S84 geographic coordinates into GCJ-02 coordinates, otherwise the obtained address will have a misplaced result.

Reply:
It probably makes sense to have some utility class for this.

Reply:
No problem, I can write a tool class for converting WG-S84 to GCJ-02 geographic coordinates in the helper package. What do you think about this offer?

Reply:
You mean your PR in general. Makes sense to support this option.